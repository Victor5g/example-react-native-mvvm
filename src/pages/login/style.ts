import { StyleSheet } from "react-native";

const style = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    paddingleft: 40,
    paddingRight: 40,
  },

  input: {
    borderColor: "#000",
    borderWidth: 0.3,
    borderRadius: 6,
    width: 230,
    height: 35,
    marginBottom: 50,
    fontSize: 16,
    alignItems: "center",
    justifyContent: "center",
    placeholderTextColor: "#808080",
    paddingLeft: 8,
    paddingRight: 8,
  },

  button: {
    width: 140,
    height: 38,
    backgroundColor: "#000",
    borderRadius: 10,
    borderWidth: 1,
    alignItems: "center",
    justifyContent: "center",
  },

  buttonTitle: {
    color: "#FFFF",
    fontSize: 23,
  },
});

export default style;
