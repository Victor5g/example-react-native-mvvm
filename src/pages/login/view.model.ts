import { useState } from "react";
import { Alert } from "react-native";
import { login } from "../../repositories/auth.repository";

const useLoginViewModel = () => {
  const [email, setEmail] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const onSubmit = async () => {
    console.log("Log: ", { email, password });
    try {
      setIsLoading(true);
      const response = await login({ email, password });
      console.log("Response:", response);
    } catch (error) {
      Alert.alert("Oops!!,", error);
    } finally {
      setIsLoading(false);
    }
  };

  return {
    email,
    password,
    isLoading,
    setEmail,
    setPassword,
    onSubmit,
  };
};

export default useLoginViewModel;
