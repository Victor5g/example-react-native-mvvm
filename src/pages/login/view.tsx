import React from "react";
import {
  SafeAreaView,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from "react-native";

import useLoginViewModel from "./view.model";

import style from "./style";

const LoginView: React.FC = () => {
  
  const { email, password, setEmail, setPassword, isLoading, onSubmit } = useLoginViewModel();

  return (
    <SafeAreaView style={style.container}>
      <View>
        <Text>E-mail</Text>
        <TextInput
          style={style.input}
          placeholderTextColor={style.input.placeholderTextColor}
          placeholder="example.email@gmail.com"
          value={email}
          onChangeText={setEmail}
        />

        <Text>Password</Text>
        <TextInput
          style={style.input}
          placeholderTextColor={style.input.placeholderTextColor}
          placeholder="********"
          value={password}
          onChangeText={setPassword}
        />
      </View>

      <TouchableOpacity
        style={style.button}
        onPress={onSubmit}
        disabled={isLoading}
      >
        <Text style={style.buttonTitle}>{isLoading ? "...." : "login"}</Text>
      </TouchableOpacity>
    </SafeAreaView>
  );
};

export default LoginView;
