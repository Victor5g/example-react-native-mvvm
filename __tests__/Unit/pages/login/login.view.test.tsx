import renderer from "react-test-renderer";
import { LoginViewModel } from "../../../../src/pages/login/model";
import LoginView from "../../../../src/pages/login/view";

//Mock Hook-Use
jest.mock("../../../../src/pages/login/view.model", () => {
  const fnLoginViewModel = jest.fn();
  fnLoginViewModel.mockImplementation(
    () =>
      ({
        email: "",
        password: "",
        isLoading: false,
        setEmail: jest.fn(),
        setPassword: jest.fn(),
        onSubmit: jest.fn(),
      } as LoginViewModel)
  );

  return fnLoginViewModel;
});

import useLoginViewModel from "../../../../src/pages/login/view.model";

describe("Unit-Test: <LoginView/>", () => {
  it("Renders correctly", () => {
    const { toJSON } = renderer.create(<LoginView />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("Show correctly with E-mail", () => {
    useLoginViewModel.mockImplementation(
      () =>
        ({
          email: "test-jest@unit.com",
          password: "",
          isLoading: false,
          setEmail: jest.fn(),
          setPassword: jest.fn(),
          onSubmit: jest.fn(),
        } as LoginViewModel)
    );
    const { toJSON } = renderer.create(<LoginView />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("Show correctly with Password", () => {
    useLoginViewModel.mockImplementation(
      () =>
        ({
          email: "",
          password: "jest@345.98",
          isLoading: false,
          setEmail: jest.fn(),
          setPassword: jest.fn(),
          onSubmit: jest.fn(),
        } as LoginViewModel)
    );
    const { toJSON } = renderer.create(<LoginView />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("Show correctly with E-mail and Password", () => {
    useLoginViewModel.mockImplementation(
      () =>
        ({
          email: "test-jest@unit.com",
          password: "jest@345.98",
          isLoading: false,
          setEmail: jest.fn(),
          setPassword: jest.fn(),
          onSubmit: jest.fn(),
        } as LoginViewModel)
    );
    const { toJSON } = renderer.create(<LoginView />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("Show correctly while is Loading", () => {
    useLoginViewModel.mockImplementation(
      () =>
        ({
          email: "test-jest@unit.com",
          password: "jest@345.98",
          isLoading: true,
          setEmail: jest.fn(),
          setPassword: jest.fn(),
          onSubmit: jest.fn(),
        } as LoginViewModel)
    );
    const { toJSON } = renderer.create(<LoginView />);
    expect(toJSON()).toMatchSnapshot();
  });
});
