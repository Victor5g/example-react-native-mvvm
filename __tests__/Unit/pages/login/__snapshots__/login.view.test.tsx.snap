// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit-Test: <LoginView/> Renders correctly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "alignContent": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 20,
    }
  }
>
  <View>
    <Text>
      E-mail
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="example.email@gmail.com"
      value=""
    />
    <Text>
      Password
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="********"
      value=""
    />
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        login
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`Unit-Test: <LoginView/> Show correctly while is Loading 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "alignContent": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 20,
    }
  }
>
  <View>
    <Text>
      E-mail
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="example.email@gmail.com"
      value="test-jest@unit.com"
    />
    <Text>
      Password
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="********"
      value="jest@345.98"
    />
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
          Object {},
        ]
      }
    >
      <Text
        disabled={true}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            Object {
              "color": "#cdcdcd",
            },
          ]
        }
      >
        ...
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`Unit-Test: <LoginView/> Show correctly with E-mail 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "alignContent": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 20,
    }
  }
>
  <View>
    <Text>
      E-mail
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="example.email@gmail.com"
      value="test-jest@unit.com"
    />
    <Text>
      Password
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="********"
      value=""
    />
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        login
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`Unit-Test: <LoginView/> Show correctly with E-mail and Password 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "alignContent": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 20,
    }
  }
>
  <View>
    <Text>
      E-mail
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="example.email@gmail.com"
      value="test-jest@unit.com"
    />
    <Text>
      Password
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="********"
      value="jest@345.98"
    />
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        login
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`Unit-Test: <LoginView/> Show correctly with Password 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "alignContent": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 20,
    }
  }
>
  <View>
    <Text>
      E-mail
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="example.email@gmail.com"
      value=""
    />
    <Text>
      Password
    </Text>
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="********"
      value="jest@345.98"
    />
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        login
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
